### ISSUE TITLE
Add cmd_mkdir configuration option

### ISSUE THREAD

**Issue by mb5** [rsnapshot/rsnapshot#20]
For modern file systems, I would be useful to have a custom mkdir command for non-temporary directories, eg.
```
cmd_mkdir=/sbin/btrfs subvolume create
```

**Comment by bebehei** [comment]
Can you explain please, why this is necessary for you?

If you look at the code, you'll see the only mkdir-calls are perl-routines. Changing now this to a shell command, it'll give only extra hassle.

**Comment by mb5** [comment]
As I said, I am using btrfs snapshots with rsnapshot.
This entails that directories have to be subvolumes, which are created by the command shown above.
Not having this option means that one needs to take special care of creating direcotries manually, which requires you to be extra careful when using options such as sync_first because it requires you to create a .sync subvolume by hand that is not even documented anywhere (to my knowledge).

**Comment by bebehei** [comment]
The .sync is just a directory. BTRFS - and in general a UNIX FS - should also support mkdir-calls.

AFAIK creating a subvolume is just smth. like mounting another physical block-device. TBH I have almost no knowledge about BTRFS, but I know, it should be supported out of the box by rsnapshot.

**Comment by mb5** [comment]
@bebehei Of course btrfs supports mkdir, but mkdir creates a directory, not a subvolume.
Contrary to what you might have heard, a subvolume has nothing at all to do with block devices.
A subvolume is in fact more of a special type of directory that btrfs knows about and that has to be created using a special command. The main difference in this case is that a subvolume can use snapshots, which basically means "create an additional filesystem pointer to the current state of that directory". This is way, way more efficient and faster than using hardlinks and cp/rsync as snapshots are atomic operations that do not even take milliseconds.

I am not sure why you support replacing cp and rsync, but are not willing to support replacing mkdir if it makes a superior backup strategy easier for users.

**Comment by bebehei** [comment]
Pls send logs and information to reproduce and I will look over it.

**Comment by mb5** [comment]
@bebehei 
Gladly, here you go: https://gist.github.com/mb5/51e126b26542994ffa45

**Comment by bebehei** [comment]
shack-mitglied?

**Comment by mb5** [comment]
Sorry, don't know what you mean.
Can you elaborate?

**Comment by nioncode** [comment]
I would also like to see this implemented.
Is there any progress?

**Comment by bebehei** [comment]
No, I started digging into it a few months ago, but I didn't get any clue, why it fails.

**Comment by nioncode** [comment]
The issue is not, that rsnapshot fails by default, because it doesn't.

Problems arise, when you try to use btrfs snapshots instead of the default 'cp' command.
The advantage of using btrfs snapshots is that the copy will be instantly. In contrast, invoking cp with hardlinks takes a few minutes and I don't even have that much files in my backup at the moment.

Why btrfs snapshots don't work by default: you can only snapshot an existing btrfs subvolume.
When you start a rsnapshot backup (with syncfirst option enabled), it creates the .sync folder using the mkdir command, which does not know anything about btrfs subvolumes.
When you then copy the .sync folder to hourly.0, the btrfs snapshot fails, because the .sync folder is not a subvolume.

To fix this, you have to manually create a .sync subvolume for all your backup destinations before invoking rsnapshot.
Therefore, it would be great if we could modify the command that gets used to create the directory ('mkdir') to use the btrfs subvolume creation ('btrfs subvolume create').

It is already possible to alter the commands for cp, rm, rsync, ssh, logger, du, rsnapshot-diff. So why don't we introduce an cmd_mkdir option? This would solve all issues.

**Comment by nioncode** [comment]
I just looked through the source code and it seems like the only place where a change would be needed is in create_backup_point_dir().
Instead of using mkpath("$destpath", 0, 0755), we would either call 'mkdir -m 0755 -p $destpath/' like the log statements says or 'btrfs subvolume create $destpath'.

**Comment by bebehei** [comment]
Interesting. My research about BTRFS in combination with rsnapshot had been completely different. rsnapshot executed without any error, but there was no snapshot in general.

If you could create a PR, this would be great.

**Comment by nioncode** [comment]
Unfortunately, I currently don't have enough time to test this, because my backup server does not use btrfs.
I might be able to create a PR, but testing it with all possible options (like --link-dest) would take a while.

**Comment by bhy** [comment]
A guide on using rsnapshot with btrfs: http://it.werther-web.de/2011/10/23/migrate-rsnapshot-based-backup-to-btrfs-snapshots/

**Comment by mf69** [comment]
This is not a bug, it's a feature request to make rsnapshot support subvolumes by nature.
As it stands I've been using btrfs with rsnapshot "normally" and it works just fine.

**Comment by mb5** [comment]
@mf69 There was never any dispute over that, albeit the classification of "bug", which is really just classification as it fits into the workflow of the development team, so no point discussing it.

@bebehei any chance to get @nioncode's pull request merged anytime soon?
